File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int n = nums.length;
        int i = 0, j = 0;
        int count = 0, maxlen = 0;
        while(j<n){
            if(nums[j]==0) count++;
            if(count<=k){
                maxlen = Math.max(maxlen,j-i+1);
            }
            while(count>k){
                if(nums[i]==0) count--;
                i++;            
            }
            maxlen = Math.max(maxlen,j-i+1);
            j++;
        }
        return maxlen;
    }
}

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int i=0;
        int j=0;
        int n=nums.size();
        int count=0;
        int zero=0;
        int ln=INT_MIN;
        while(j<n){
            if(nums[j]==1) count++;
            else zero++;
            if(zero<=k){
                ln=max(ln,j-i+1);
            }
            while(zero>k){
               if(nums[i]==1) {
                count--;
                i++;
               }
                else{
                    zero--;
                    i++;
                    break;
                }
            }
            j++;
        }
        if(ln==INT_MIN) return 0;
        return ln;
    }
};